mixnet/protocol/src/lib.rs

```diff
@@ -99,3 +104,39 @@ impl Body {
         Ok(())
     }
 }
+
+pub async fn retry_backoff(
+    peer_addr: SocketAddr,
+    max_retries: usize,
+    retry_delay: Duration,
+    body: Body,
+    socket: Arc<Mutex<TcpStream>>,
+) -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
+    for idx in 0..max_retries {
+        // backoff
+        let wait = retry_delay * (idx as u32);
+
+        tokio::time::sleep(wait).await;
+        let mut socket = socket.lock().await;
+        match body.write(&mut *socket).await {
+            Ok(_) => return Ok(()),
+            Err(e) => {
+                if let Some(err) = e.downcast_ref::<std::io::Error>() {
+                    match err.kind() {
+                        ErrorKind::Unsupported
+                        | ErrorKind::NotFound
+                        | ErrorKind::PermissionDenied
+                        | ErrorKind::Other => {}
+                        _ => {
+                            // update the connection
+                            if let Ok(tcp) = TcpStream::connect(peer_addr).await {
```

Comment:
what's the rational behind updating the connection upon receiving these errors? For example, `Unsupported` means the action can never succeed, why retrying doing the same thing as before?

Reply:
I am a little bit confused here. It seems that the code will not retry on `Unsupported`.
```
 ErrorKind::Unsupported
                        | ErrorKind::NotFound
                        | ErrorKind::PermissionDenied
                        | ErrorKind::Other => {},
```

Reply:
I am a little bit confused here. It seems that the code will not update the conncection on those errors, do you mean we should early return when those errors happen?
```rust
 ErrorKind::Unsupported
                        | ErrorKind::NotFound
                        | ErrorKind::PermissionDenied
                        | ErrorKind::Other => {}
```

Reply:
The fact that `ErrorKind::Unsupported` is specifically mentioned suggests we have some specific appropriate handling for it and due to the nature of the error it would mean to stop retying.
It's acceptable to retry again if we don't want to differentiate errors, but if we differentiate then it only makes sense to do the right thing.

Reply:
Now we return early for `ErrorKind::Other`, why is that?

Reply:
Ah, I misunderstood. so we only need to early return for `Unsupported` here.

Reply:
My point was that if we explicitly mention an `ErrorKind` then we need to appropriately handle it. Depending on the specific case, it could be that `Other` and `Unsupported` might be treated differently (like this one).

Reply:
Make sense. 👍